### PR TITLE
Treat deployment with empty image as paused

### PR DESCRIPTION
**Pull request by mfojtik**
**What this PR does / why we need it**:

When a deployment have empty image specified either in Containers or InitContainers, the rollout of such deployment is likely going to fail or be incomplete. This change will treat such deployment as it was paused, which allows external processes to supply an image and only after then trigger the initial rollout.

@smarterclayton we need this to make the `new-app` flow working for deployments.

TODO: 

- [x] Need test for empty images and for pause
- [x] Need an event to tell users reason why the deployment is paused

**Release note**:
```release-note
Deployments with empty image specified in any container are now acting as "paused" deployments.
```

### CODE REVIEW

**Comment by mfojtik** [comment]
I think we can also update the deployment and set pause=true/false based on the presence of the image, but this seemed like a smaller change (no need to mutate the deployment and resync)

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -556,6 +557,21 @@ func (dc *DeploymentController) getPodMapForDeployment(d *extensions.Deployment,
 	return podMap, nil
 }
 
+// hasEmptyImage returns true if any of the deployment containers have empty image in the container spec.
+func hasEmptyImage(d *extensions.Deployment) bool {
+	for _, container := range d.Spec.Template.Spec.Containers {
+		if len(strings.TrimSpace(container.Image)) == 0 {
```
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Do we allow empty images in Deployments?

Comment (mfojtik) [reply pkg/controller/deployment/deployment_controller.go]: @kargakis @tnozicka i believe PodSpec image can be now empty

Comment (mfojtik) [reply pkg/controller/deployment/deployment_controller.go]:
```go
	// Docker image name.
	// More info: https://kubernetes.io/docs/concepts/containers/images
	// This field is optional to allow higher level config management to default or override
	// container images in workload controllers like Deployments and StatefulSets.
	// +optional
	Image string `json:"image,omitempty" protobuf:"bytes,2,opt,name=image"`
```

Comment (mfojtik) [reply pkg/controller/deployment/deployment_controller.go]: the PodSpec validation have:

```go
// TODO: do not validate leading and trailing whitespace to preserve backward compatibility.
// for example: https://github.com/openshift/origin/issues/14659 image = " " is special token in pod template others may have done similar
if len(ctr.Image) == 0 {
  allErrs = append(allErrs, field.Required(idxPath.Child("image"), ""))
}
```

So apparently `image: " "` should pass

**Review by tnozicka — APPROVED** [review]
/lgtm

File: pkg/controller/deployment/deployment_controller.go (hunk 1 above)
Comment (tnozicka) [reply pkg/controller/deployment/deployment_controller.go]: yes, space is valid

**Comment by tnozicka** [comment]
/lgtm cancel
Actually let's have some tests for it but otherwise it look good

**Comment by tnozicka** [comment]
If there is no image specified we generally shouldn't deploy because we know for sure it's going to fail and, as you say, this is useful for external controllers injecting those images.

**Comment by mfojtik** [comment]
improved this slightly... we can't send event because this will cause a flood of events when the deployment will wait for images... so instead we should use Conditions and for that we will need a new "pending" condition I think? @kargakis @tnozicka reasonable?

**Comment by kargakis** [comment]
> improved this slightly... we can't send event because this will cause a flood of events when the deployment will wait for images...

I would like to avoid introducing a new condition. We could send an event when we haven't observed the deployment (d.metadata.generation != d.status.observedGeneration).

@kubernetes/sig-apps-pr-reviews need to chime in.

**Comment by mfojtik** [comment]
@kargakis 

> I would like to avoid introducing a new condition. We could send an event when we haven't observed the deployment (d.metadata.generation != d.status.observedGeneration).

I think the condition here is pretty natural. The deployment is "pending" an template update
and that to me is a valid condition. I would love to have a way how to reflect this state of deployment.

Event is fine, but not really visible and eventually go away. If deployment is with no image set for longer time, users will have no clue why.

FYI I added a unit test and improved the logic. I will emit event (once) and also set condition. Also I will now call updateStatus() only when the condition change.
As a bonus the even will include names of container that are awaiting images.

**Comment by kargakis** [comment]
Why the pending condition wouldn't fall under Progressing?

**Comment by dhilipkumars** [comment]
Just curious should we not validate and return an error at the time of creation ? Why should we allow this object to be created (knowing that it will fail) and then a complex workflow to maintain it ?

**Comment by mfojtik** [comment]
@kargakis progressing can work if we can set the reason and then check on that.

**Comment by mfojtik** [comment]
@dhilipkumars 
> Just curious should we not validate and return an error at the time of creation ? Why should we allow this object to be created (knowing that it will fail) and then a complex workflow to maintain it ?

An empty image (" ") allows for the flow, where you want to create the deployment but you don't have the image ready yet. In OpenShift this basically means you running a build of the image and you want to trigger the deployment when the build finish. When it finish, the ImageStream is updated and that will trigger an update for deployment image field, supplying the image it should run with.

Besides OpenShift, this can also enable flows for any external controller or CRD that is doing the image substitution (like a Jenkins pipeline or something) and you care about the "initial" deployment. 

An alternative can be that the deployment will be created with some random image (ugly) and as paused and the trigger will automatically un-pause it after it change the image to real one.

**Review by janetkuo — CHANGES_REQUESTED** [review]
This seems very hacky. Why not just pause the Deployment via `.spec.paused` (when setting image to `" "`) and unpause it when the real image is supplied?

cc @kow3ns

**Comment by mfojtik** [comment]
@janetkuo after discussing this with @kargakis we think we should not mutate the spec paused field. The fact that the deployment is "waiting" for image is a permanent condition until external entity (human, controller, etc..) does mutate the container image. This should be reflected as "progressing=false" (because the deployment is not doing any progress) and also "reason=WaitingForImage" (so user know that we wait for him to supply images).

If we just mutate the spec.paused, users (or any external system) will not know the reason why we did that and mutating spec might lead into races. I think in general we are trying to avoid any spec mutations by controllers?.

I reworked this PR to not add new API condition (so this does not need an API review?) and I used the progressing=false/reason=WaitForImages as a marker for emitting the event (so we fire it up just once). Updated tests as well.

PTAL.

**Comment by tnozicka** [comment]
@janetkuo 
> This seems very hacky. Why not just pause the Deployment via .spec.paused (when setting image to " ") and unpause it when the real image is supplied?

What if you want to have a paused Deployment that would got it's images injected and you want to unpause it manually at specific time? How would the controller know when to unpause it and when not while injecting the image?

Also this way, if user would forget to pause the Deployment, it would also deploy an RC with empty image.
I see this as issue of correctness - the controller shouldn't deploy RCs that it knows for sure have zero chance of getting deployed.

**Comment by mfojtik** [comment]
@tnozicka agree, if user un-pause deployment manually, then it will get deployed and the RS will be created with image = " " which will result in failed rollout.

File: pkg/controller/deployment/util/deployment_util.go [hunk 2]
```diff
@@ -91,6 +91,10 @@ const (
 	// ResumedDeployReason is added in a deployment when it is resumed. Useful for not failing accidentally
 	// deployments that paused amidst a rollout and are bounded by a deadline.
 	ResumedDeployReason = "DeploymentResumed"
+	// WaitingForImagesReason is added in a deployment when the deployment has one or more image fields empty either in
+	// initContainers or containers. In that case the progress of this deployment is false and we have to wait until
+	// user or external entity provide valid, non-empty image fields.
+	WaitingForImagesReason = "WaitingForImagesReason"
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: Remove Reason from the value.

File: pkg/controller/deployment/deployment_controller.go [hunk 3]
```diff
@@ -556,6 +557,44 @@ func (dc *DeploymentController) getPodMapForDeployment(d *extensions.Deployment,
...
+		return true
+	}
+	if isWaiting {
+		util.RemoveDeploymentCondition(&d.Status, extensions.DeploymentProgressing)
+		// If error occurs during updating status, don't emit the event but reconcile one more time.
+		if _, err := dc.client.ExtensionsV1beta1().Deployments(d.Namespace).UpdateStatus(d); err == nil {
+			// Only fire event once and only when the update succeed.
+			dc.eventRecorder.Eventf(d, v1.EventTypeNormal, util.WaitingForImagesReason, "This deployment has now all image fields set.")
+		}
+		// At this point we already have the images available, but let make controller reconcile this again
```
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Shouldn't you force a resync until the cached object is updated?

Comment (kargakis) [reply pkg/controller/deployment/deployment_controller.go]: Oh, well, once the cached object is updated we get an event, right?

**Comment by kargakis** [comment]
> This should be reflected as "progressing=false" (because the deployment is not doing any progress) and also "reason=WaitingForImage" (so user know that we wait for him to supply images).

I don't feel strong on whether this should be "status=false" or "status=unknown", but I think we should avoid mutating spec.

File: pkg/controller/deployment/deployment_controller.go (hunk 3 above)
Comment (mfojtik) [reply pkg/controller/deployment/deployment_controller.go]: right

**Comment by mfojtik** [comment]
@kargakis i like "False" more as the deployment is really not taking any progress when it waits for the images to be updated. The "Unknown" might suggests that controller can't determine the state or is in some error/weird state about the deployment.

File: pkg/controller/deployment/deployment_controller.go (hunk 3 above)
Comment (mfojtik) [reply pkg/controller/deployment/deployment_controller.go]: (also this rely on the condition/reason to be set, so unless that is not reflected in cache, this will keep retrying updating the condition. once it succeed it will emit the event once).

**Comment by kargakis** [comment]
This feature is specifically to enable higher level orchestrators to resolve the image and allow progress so while the controller has no view of Deployment progress, it cannot know about other processes hence I think "Unknown" is more fit. Not something I am crazy about though.

**Comment by mfojtik** [comment]
@kargakis ok, changed to Unknown, don't care that much :-)

**Comment by janetkuo** [comment]
> we need this to make the `new-app` flow working for deployments.

@mfojtik would you explain what you plan to achieve with this change? And why can't this logic live outside of core? This would make it easier to review this change.

**Comment by mfojtik** [comment]
@janetkuo in OpenShift there is concept of [new-app command](https://docs.openshift.org/latest/dev_guide/application_lifecycle/new_app.html#using-the-cli) that is used to create "flows".
It takes the source repository (GIT/etc.) and the builder image name as an input and it generates:

1) ImageStream for application image
2) BuildConfig definition of how the application image should be build
3) DeploymentConfig (we want to enable Deployment here, yay!) that deploys the application

All these resources are create when the `new-app` command runs. We don't wait for builds to complete.

DeploymentConfig has an [image change trigger](https://docs.openshift.org/latest/dev_guide/deployments/basic_deployment_operations.html#image-change-trigger), that will kick the initial rollout only when the ImageStream for application image is modified (image is available / pushed to registry)... 

The problem that I have with adapting Deployments to this flow is that Deployments does not really have the concept of "wait for the image to be available".
The closes way to achieve this is to create a "paused" Deployment. That puts the burden on the trigger mechanism, where the trigger has to decide when to resume the Deployment for initial rollout. 
Trigger, however, has no way to know why the Deployment is paused (should it always resume? what is initial rollout? should it rely on generation? etc).

Using the empty image field seems right to me, because in fact we don't have the image available yet, so the image is "unknown" at the time of Deployment creation.

Also I believe the change to controller is minimal (just check if images are empty and reconcile), most of the code is to handle the messaging to users.

/cc @smarterclayton

**Comment by tnozicka** [comment]
@janetkuo 

> The user (or external tool) should pause the Deployment when they set its image to " " (it's done at the same time so there's no race) if they want this behavior, and unpause it while they're ready to set the image.

It's not the controller who sets image to " " but the user when initially creating the Deployment. 

> I'd even argue that Deployment shouldn't allow " " as a valid image name.

That would force all the people using the feature of injecting images to put a made up image there for the interim before the real image could be resolved and injected.

Let me give you an example of how Deployments and ImageStreams work in Openshift to better understand the need.

```yaml
apiVersion: v1
kind: ImageStream
metadata:
  name: znc
spec:
  tags:
  - name: latest
    from:
      kind: DockerImage
      name: docker.io/tnozicka/znc:latest
    importPolicy:
      scheduled: true
```

```yaml
apiVersion: apps/v1
kind: Deployment 
metadata:
  name: znc
  annotations:
    image.openshift.io/triggers: |-
      [
        {"from":{"kind":"ImageStreamTag","name":"znc:latest"},"fieldPath":"spec.template.spec.containers[0].image"}
      ]
spec:
  template:
    spec:
      containers:
      - name: znc
        image: " "
# ...
```
The user doesn't know the image and he doesn't need to, it will get injected by a controller from the image stream.

**Comment by ironcladlou** [comment]
Seems to me like the higher level app creation orchestrator should do a "create or update" operation to handle the first-time case: if there is no image to deploy, it's nonsensical to create a deployment. After the build, when the deployment is available, create the valid deployment. When new images are built, trigger an update the existing deployment to a new valid state.

Creating a "temporarily invalid" deployment seems really fishy. I agree with @janetkuo that an empty image name is probably invalid. Sounds more like deployment templates than deployments.

**Comment by ironcladlou** [comment]
Creating a new paused or 0-replica deployment with a bogus image name seem like less invasive workarounds.

**Comment by smarterclayton** [comment]
The use case is integrating other conditions / steps into a workload controller, without the workload controllers having to have a ton of special case logic.  I think the problem with "image" as a special case is that there are other cases, like resources or environment, that we've identified and solved in other ways.  The reason for the loose coupling is to break the ordering dependency on create - just like we don't want to require users to create services after workload objects in order to get env vars, or secret refs on pods block pod startup so we don't have races.

I'd be inclined to say that this is an initialization problem - the author of the deployment knows that they don't want to do a deployment until an image exists, or until a service exists, or until another component updates a config map.  But they have no way of communicating that declaratively (end users can't set initializers).  We have crude blockers today for config maps and secrets.  Perhaps what we are really missing is a more user focused condition / initializer concept that most workload controllers should.  Something like a "pause condition" that users can set and controllers can clear:

```
type DeploymentSpec struct {
  ...
  Paused bool
  PauseConditions []PauseCondition
}

type PauseCondition struct {
  // Reason is a machine readable reason for why this is paused
  Reason string 
  // Message is a human readable message about why this is paused
  Message string
}
```

A human could set `paused: true` or `pauseConditions: [{reason: "OutageWindow", message: "Changes are prohibited over the Christmas break unless explicitly approved."}]`.  A company specific OutageController or admission webhook could add pauseConditions on all deployments labelled a certain way.  The workload controller could use pauseConditions as an additional source of delay (and an older client would get condition updates that progressing was blocked by pause conditions).

As a human, I might want to `kubectl set paused --reason PTO --message "Nobody mess with this until I get back"`.

**Comment by mfojtik** [comment]
@ironcladlou if we generate a paused deployment with 0 replicas then the trigger (annotation) has to understand that this is an initial deployment and will have to unpause it and scale (I don't think 0 replicas would be right). That means trigger will have to get smarter about how to deal with paused deployments.

@smarterclayton I understand the problem you are describing that we need a better way to communicate the reason of why the deployment is paused.
However, with the current state of API, where we do allow an empty image in PodSpec, we can either validate that in Deployments (likely break backward compatibility?) or take advantage of it. 

I agree with @ironcladlou that this looks more like a "deployment template" problem, where you created a deployment but want to substitute the image later, because you don't know what the image will be (an external process will supply it).

I would even say that the current deployment flow is broken as if you specify the empty image for a not-paused deployment, it will result in broken rollout. If we pause such deployment automatically, we will prevent this rollout from happening and also give external entity time to fix the deployment image.

I think the changes proposed in this PR are still valid, relatively minimal and fix the broken rollout problem. WDYT?

**Comment by janetkuo** [comment]
A pod's container image must not have leading or trailing whitespace:

https://github.com/kubernetes/kubernetes/blob/884e4b7b9c63a9a3b94278972d640454b8d16e47/pkg/apis/core/validation/validation.go#L2800-L2811

Pod creation would fail if this is violated. Therefore, one cannot create a pod with its image set to `" "`. The reason that we didn't do this validation at pod template level (i.e. in deployments or replicasets) is to preserve backward compatibility.

**Comment by mfojtik** [comment]
@janetkuo i thought that the fact we already do support `" "` as image in deployments and replicasets (even if that is for backward compatibility) we can take advantage of that and use it to support the:
"I have deployment that is currently waiting for the image to be provided".

I don't think that deployment is invalid, it just have nothing to deploy (yet). In that case I still think "pausing" the rollout till it has something to deploy is valid case or at least fixes the current problem of rolling out replica sets that are not able to create pods.

**Comment by dhilipkumars** [comment]
@mfojtik && @tnozicka I agree with others,  we shouldn't be building new functionalities on top of things that we are trying to get rid off.  Plus we should replicate this behavior to other controllers aswell. 

As a workaround why cant the CI system use [`pause`](https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/rollingupdate-statefulset-rv2.yaml#L29) container for those deployments whose images are not ready yet and then update them suitably when they are ready?

**Comment by mfojtik** [comment]
@dhilipkumars how the CI system is going to distinguish between a "paused" deployment because it waits for images vs. a "paused" deployment because user paused it manually?

**Comment by mfojtik** [comment]
@smarterclayton another solution might be:

1) generate as "paused" deployment and set "resume-on" annotation to "image-available"
2) then make trigger controller resume this deployment when it injects image field and clear that annotation.

**Comment by tnozicka** [comment]
As it has been said many times here, **we can't change the validation as that is breaking change**, nor I want to as it breaks integrations that rely on it to inject image after the deployment is created. 

So we are clearly left with the state where user can specify such image and we know for sure it'll fail. **Creating a RS which will fail with absolute certainty looks like a bug to me**.

Yeah, we can probably hack our way around it outside of Kubernetes by pausing it in admission and trying to figure out who paused and if we should unpause it, but none of it will change the fact that Deployment controller deploys invalid RS. So I'd like us to collaborate here and fix it the two issues together if possible.

@janetkuo Is it the extent of the change that discourages you? Would it be more acceptable if we went back to the simple fix that it has been in the beginning? Just check if the image is empty and don't create RS? It can get down to be just 3-5 LOC not to impose maintenance burden.

**Comment by tnozicka** [comment]
@kow3ns any opinions about this whole thread?

**Comment by krmayankk** [comment]
@mfojtik @janetkuo @kargakis @smarterclayton why should we give special treatment to image being empty in deployment ? for all workload apis, image not being present should be treated as error and fail rollout of that workload. If we want to pause in deployment, what should be the behavior in statefulsets, or RS or Pods , or Jobs, etc ? IMO the behavior should be consistent across all workload apis

**Comment by janetkuo** [comment]
@fejta this is already closed, but the bot is still spamming it

**Comment by BenTheElder** [comment]
See: https://github.com/kubernetes/test-infra/issues/9001 :/

**Comment by BenTheElder** [comment]
TLDR; the github search API is broken and causing issues like this, they are working on it.

**Comment by BenTheElder** [comment]
The bot should be disabled now, (unfortunately not before the comment above), we will re-enable it after the GitHub API is fixed. thanks for reporting @janetkuo
